### PR TITLE
refactor: Add full pair iterator and for_each_full_pair

### PR DESCRIPTION
**Pull request by rprospero**
This adds a `FullPairIterator` class and a `for_each_full_pair` algorithm that examines each ordered pair (as opposed to `for_each_pair`, which only examines unique unordered pairs).

### CODE REVIEW

File: src/classes/fullPairIterator.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+bool FullPairIterator::operator==(const FullPairIterator &other) const { return x_ == other.x_ && y_ == other.y_; }
+
+bool FullPairIterator::operator!=(const FullPairIterator &other) const { return x_ != other.x_ || y_ != other.y_; }
+
+FullPairIterator::value_type FullPairIterator::operator*() { return {x_, y_}; }
+
+FullPairIterator &FullPairIterator::operator++()
+{
+    y_++;
+    if (y_ >= size_)
```
Comment (trisyoungs) [inline src/classes/fullPairIterator.cpp]:
```suggestion
    if (++y_ >= size_)
```

Comment (trisyoungs) [reply src/classes/fullPairIterator.cpp]: Just a suggestion!

File: tests/algorithms/pairIterator.cpp [hunk 2]
```diff
@@ -23,4 +24,20 @@ TEST(PairIterTest, Pairs)
...
+    {
+        for (y = 0; y < size; ++y)
+        {
+            auto [i, j] = *it;
+            EXPECT_EQ(i, x);
+            EXPECT_EQ(j, y);
+            ++it;
+        }
+    }
+}
```
Comment (trisyoungs) [inline tests/algorithms/pairIterator.cpp]: I want to suggest here that we do a less "pathological" test of the new iterator. Specifically something like taking a full (i.e. not halved) `Array2D` of random ints and get the sum through both the dumb double x,y loop as well as via the new iterator.

Comment (rprospero) [reply tests/algorithms/pairIterator.cpp]: I had a whole argument for why we shouldn't do this, but now I think that you're right.  My argument had been that using the sum test proves that every pair is visited, but doesn't prove that we visited them in the correct order.  However, it then becomes a philospohical question of whether there *is* a correct order or if that should just be considered an implementation detail.  I'm now leaning closer to making it an implementation detail, so none of the tests should depend on it.

**Review by trisyoungs — APPROVED** [review]
Think this all looks good, although I'm a bit foggy!

**Pull request merged**